src/liballoc/collections/btree/node.rs

```diff
@@ -213,27 +194,31 @@ impl<K, V> Root<K, V> {
         ret
     }
 
-    /// Removes the root node, using its first child as the new root. This cannot be called when
-    /// the tree consists only of a leaf node. As it is intended only to be called when the root
-    /// has only one edge, no cleanup is done on any of the other children of the root.
-    /// This decreases the height by 1 and is the opposite of `push_level`.
-    pub fn pop_level(&mut self) {
-        assert!(self.height > 0);
-
-        let top = self.node.ptr;
+    /// Removes the root node if has a single child thus length zero, using that
+    /// child as the new root, and returning `true`. This decreases the height
+    /// by 1 and is the opposite of `push_level`. If the root node has no children
+    /// because it is a leaf, or if it has multiple children because length is
+    /// non-zero, changes nothing and returns `false`.
+    pub fn pop_empty_level(&mut self) -> bool {
```

Comment:
Could we rename this to e.g. `pop_level_if_empty` or so? I get confused seeing it otherwise where the level is not necessarily empty.

Reply:
I admit it's not a consistent name, but to be clear: if it gets popped, it was an empty level. Like LinkedList::pop_front, that pops if there is a front and does nothing if there isn't. But here... also not if it's an empty leaf. The accurate name would be a long `pop_level_if_empty_and_internal`. Which I don't mind at all, but in general people seem to dislike descriptive names.

On the other hand, the name of Vec::pop doesn't give any clue about when it would refuse to pop (because it's whenever possible). For my part, the name `pop_level` could stay; go read the documentation if you wonder when it will operate.

At the moment, I think `pop_empty_internal_level` would be best.

Reply:
LinkedList::pop_front (and similar methods on Vec, etc.) *always* pop if there's any element to be had. I think some sort of "if" here is needed, and I'd be quite happy with pop_level_if_empty_and_internal personally.

The problem with names omitting the if -- at least for me -- is that they rather sound like assertions, i.e., "hey only call this if you're popping an empty and internal level" whereas the code in this function in fact works regardless -- it'll check if that's the case itself.

Reply:
Wait a second... apart from a few (mostly recent) comments, nothing says that a leaf is a level. If height is zero, is it a level? `push_level` always pushes an internal node, so surely `pop_level` will only pop the same kinds of thing. So `pop_level_if_empty` would be perfectly unambiguous. Well, it would be nicer to offer `push_internal_level` and `pop_internal_level_if_empty`.

 The decision to keep an empty leaf is something that should remain at the map level, so maybe I should add a function at the map level, and just

Reply:
Adding internal makes sense to me. I personally don't like calling leaves not a level of their own -- seems unintuitive to me.